### PR TITLE
fix(learn): Update text as code for vector-dot-product

### LINKED ISSUE #40036: The term "nVectors" in "Rosetta Code: Vector dot product" is undefined
**Describe your problem and how to reproduce it:**
One of the test cases on the sidebar says `dotProduct(...nVectors) should return 156000`.  nVectors is undefined

Additionally, I notice that there are some tests where the text value doesn't match the actual test string.

**Add a Link to the page with the problem:**
https://www.freecodecamp.org/learn/coding-interview-prep/rosetta-code/vector-dot-product

**Tell us about your browser and operating system:**
* Browser Name: Firefox
* Browser Version: 82
* Operating System: Linux Mint

### PR DESCRIPTION
**Pull request by RandellDawson**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

Closes #40036

This is a replacement PR for #40168.  When I resolved a merge conflict on that PR, somehow the PR ended up with 0 commits and the PR auto-closed.  I added all the suggested changes for the challenge and converted to the new challenge format, so I will not be able to review this one now.

### CODE REVIEW

**Comment by RandellDawson** [comment]
I changed the return value for the last test from `384` to `360` which is the correct dot product of `[ 0, 3, 6, 9, 12 ]` and `[ 0, 4, 8, 12, 16 ]`.

File: curriculum/challenges/english/10-coding-interview-prep/rosetta-code/vector-dot-product.md [hunk 1]
```diff
@@ -7,57 +7,48 @@ forumTopicId: 302343
...
 
-dotProduct() should return null.
+`dotProduct()` should return `null`.
 
 ```js
 assert.equal(dotProduct(), null);
 ```
 
-dotProduct(\[[1], [1]]) should return 1.
+`dotProduct(\[[1], [1]])` should return `1`.
```
Comment (ShaunSHamilton) [inline curriculum/challenges/english/10-coding-interview-prep/rosetta-code/vector-dot-product.md:30]:
```suggestion
`dotProduct([1], [1])` should return `1`.
```

File: curriculum/challenges/english/10-coding-interview-prep/rosetta-code/vector-dot-product.md [hunk 2]
```diff
@@ -7,57 +7,48 @@ forumTopicId: 302343
...
 
 ```js
 assert.equal(dotProduct([1], [1]), 1);
 ```
 
-dotProduct(\[[1], [1, 2]]) should return null.
+`dotProduct(\[[1], [1, 2]])` should return `null`.
```
Comment (ShaunSHamilton) [inline curriculum/challenges/english/10-coding-interview-prep/rosetta-code/vector-dot-product.md:36]:
```suggestion
`dotProduct([[1], [1, 2]])` should return `null`.
```

File: curriculum/challenges/english/10-coding-interview-prep/rosetta-code/vector-dot-product.md [hunk 3]
```diff
@@ -7,57 +7,48 @@ forumTopicId: 302343
...
 ```js
 assert.equal(dotProduct([1], [1, 2]), null);
 ```
 
-dotProduct([1, 3, -5], [4, -2, -1]) should return 3.
+`dotProduct([1, 3, -5], [4, -2, -1])` should return `3`.
 
 ```js
 assert.equal(dotProduct([1, 3, -5], [4, -2, -1]), 3);
 ```
```
Comment (ShaunSHamilton) [inline curriculum/challenges/english/10-coding-interview-prep/rosetta-code/vector-dot-product.md:47]:
````suggestion

`dotProduct([3, 2, 1], [2, 4, 2], [5, 3, 1])` should return `null`.

```js
assert.equal(dotProduct([3, 2, 1], [2, 4, 2], [5, 3, 1]), null);
```

````

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
Ideally, we would add another test - which will break the solution.

Should we also change the solution code? It passes, but technically it does not meet the requirements.

File: curriculum/challenges/english/10-coding-interview-prep/rosetta-code/vector-dot-product.md (hunk 3 above)
Comment (RandellDawson) [reply curriculum/challenges/english/10-coding-interview-prep/rosetta-code/vector-dot-product.md:47]: @ShaunSHamilton Was the change part of the original PR?

Comment (ShaunSHamilton) [reply curriculum/challenges/english/10-coding-interview-prep/rosetta-code/vector-dot-product.md:47]: This specifically, I do not remember so. However, we are not properly testing the user requirement to return `null` on any input other than 2 equally-sized vectors.

Comment (RandellDawson) [reply curriculum/challenges/english/10-coding-interview-prep/rosetta-code/vector-dot-product.md:47]: True.  I will commit your suggestion.

**Comment by ShaunSHamilton** [comment]
```
function dotProduct(...vectors) {
  if (!vectors || !vectors.length || vectors.length > 2 || vectors[0].length !== vectors[1].length) {
    return null;
  }
  const vectorLen = vectors[0].length;

  let prod = 0;
  let sum = 0;
  let j = vectorLen;
  let i = 2;
  // Sum terms
  while (j--) {
    i = 2;
    prod = 1;

    while (i--) {
      prod *= vectors[i][j];
    }
    sum += prod;
  }
  return sum;
}
```

This is a quick change of the current solution to pass.

**Review by ShaunSHamilton — DISMISSED** [review]
This LGTM 👍 . I am happy with this, and believe is clarifies the requirements, and resolves the issue.

**Review by naomi-lgbt — DISMISSED** [review]
LGTM~! 🥳

I pulled this down locally, tested against the [official Rosetta Code solution](http://www.rosettacode.org/wiki/Vector_products#JavaScript), and verified everything passes. The one exception was our three array return, as the original Rosetta Code problem expects the function to handle any `n` number of arrays.

I'm good with merging this once the freeze is lifted. 🙂 

Thank you for your pull request! We look forward to additional contributions!

**Comment by ylevental** [comment]
I was the one that filed the original issue at #40036.  While the replacement code looks good, I still notice that one of the lines in the replacement code has an extra pair of brackets.  This is the top line of the code that I copied:

RandellDawson:fix/update-text-and-tests-for-vector-dot-product-challenge

```
dotProduct([[1], [1, 2]]) should return null.
assert.equal(dotProduct([1], [1, 2]), null);
```

**Comment by ShaunSHamilton** [comment]
@ylevental That is one of the things this PR already fixes.

**Comment by ylevental** [comment]
@ShaunSHamilton I also checked the "Files Changed" tab of this request: https://github.com/freeCodeCamp/freeCodeCamp/pull/40460/files

On line 36, I still see the extra pair of brackets on the green line.  Am I looking in the right place?

**Comment by ShaunSHamilton** [comment]
@ylevental Ah...I did miss that. You are quite right. Please go ahead and make a suggestion using the comment feature in GitHub, with the correct code. 👍

**Comment by ylevental** [comment]
Sure, on line 36 of the file,

``dotProduct([[1], [1, 2]])` should return `null`.` should be changed to

``dotProduct([1], [1, 2])` should return `null`.`

**Comment by ShaunSHamilton** [comment]
@ylevental I mean, go to the files tab of this PR, click on the ➕ next to the line you want to add a suggestion to, and click on the option to _Add Suggestion_ (looks like a file with a plus sign): 
image

**Comment by ylevental** [comment]
@ShaunSHamilton I see, I just did that

**Review by ylevental — DISMISSED** [review]

File: curriculum/challenges/english/10-coding-interview-prep/rosetta-code/vector-dot-product.md [hunk 4]
```diff
@@ -7,57 +7,54 @@ forumTopicId: 302343
...
 
-dotProduct(\[[1], [1]]) should return 1.
+`dotProduct([1], [1])` should return `1`.
 
 ```js
 assert.equal(dotProduct([1], [1]), 1);
 ```
 
-dotProduct(\[[1], [1, 2]]) should return null.
+`dotProduct([[1], [1, 2]])` should return `null`.
```
Comment (ylevental) [inline curriculum/challenges/english/10-coding-interview-prep/rosetta-code/vector-dot-product.md:36]: There should only be one pair of brackets per vector, so this line should be changed to

`dotProduct([1], [1, 2])` should return `null`.

Comment (ylevental) [inline curriculum/challenges/english/10-coding-interview-prep/rosetta-code/vector-dot-product.md:36]: There should only be one pair of brackets per vector, so this line should be changed to

dotProduct([1], [1, 2]) should return null.

**Comment by RandellDawson** [comment]
@ShaunSHamilton @ojeytonwilliams @moT01 @nhcarrigan I think this one is ready for final review.  Let me know if anything else should be changed.

**Review by ShaunSHamilton — APPROVED** [review]
This LGTM 👍

**Review by scissorsneedfoodtoo — APPROVED** [review]
Read through the comments in #40036 and #40168, and tested locally. Everything LGTM :+1: :+1:

I think it's totally fine to limit the scope of the Rosetta Code challenges, which tend to be more open ended. Since we're calling our version of the challenge "Vector dot product", it makes sense to only expect two vectors, and to throw an error if more are passed.

**Pull request merged**